docs/usage/events.md

```diff
@@ -1,5 +1,105 @@
 # Events
 
+An {class}`~noob.event.Event` is anything that get returned from {class}`~noob.node.base.Node`,
+wrapped in a container with some metadata.
+The metadata include the event's ID, its name, when it happened, and where it happened. It is managed by
+{class}`~noob.store.EventStore` {class}`~noob.runner.base.TubeRunner`, and
+can be accessed with runner's {meth}`~noob.runner.base.TubeRunner.add_callback` method,
+like the following:
+
+```python
+from noob import Tube, SynchronousRunner
+
+tube = Tube.from_specification("...")
+runner = SynchronousRunner(tube=tube)
+
+
+def my_cb(): ...
+
+
+runner.add_callback(my_cb)
+```
+
+## EventStore
+
+The storage for all relevant events. When you're running a {class}`~noob.runner.sync.SynchronousRunner`, the
+{class}`~noob.store.EventStore` is owned by the runner who manages all nodes, so this includes every event that gets
+emitted by every node (unless it gets cleared.)
+
+On the other hand, {class}`~noob.runner.zmq.ZMQRunner` does not manage a global {class}`~noob.store.EventStore`. Rather,
+each {class}`~noob.runner.zmq.NodeRunner` manages events that are relevant to the single node it manages
+(ref: {doc}`zmq`). Either way, the functionality remains the same:
+preserving relevant events while there remains a node that may depend on it,
+and collecting the values from within the events and returning them to the `runner`
+so that a relevant node can use them.
+
+## MetaEvent
+
+Remember earlier when we said Events are things that get returned from nodes? We lied. Sort of. Tee hee.
+
+We felt justifying in lying, because the other types of events are the "internal" events do not get exposed to the user.
+We call these events {class}`~noob.event.MetaEvent`. These work by swapping the `signal` entry of the event with
+{class}`~noob.event.MetaEventType`, which includes `NodeReady` and `EpochEnded`.
+Their meanings are quite self-explanatory, where `NodeReady` signals that a node's dependencies have been satisfied
+and the node is ready to process, and `EpochEnded` means... well, the given epoch (a full cycle of graph) has completed.
+For debugging purposes, we do allow users to access some of these through runner callback. This behavior may change in
+the future, however, so we do not recommend you depend on it.
+
 ## Optional Events
 
-(discuss returning `None` and `NoEvent` here)
\ No newline at end of file
+Sometimes, a node may decide not to emit anything. It accepted its inputs, processed, and nothing came of it.
```

Comment:
i can't write anymore rn but i think we'll need to go into greater detail on here and illustrate the different choices that could have been made here.